src/Evaluator.hs

```diff
@@ -310,53 +481,91 @@ doTypeCase blameLoc (Ty v0) ((p, rhs0) : ps) = match (doTypeCase blameLoc (Ty v0
         ValueSyntax (Syntax (Stx _ _ (String str))) ->
-          withExtendedEnv n x (ValueString str) (eval rhs)
+          step $ Down (unCore rhs) (extend n x (ValueString str) env) k
         _ -> next
-    match next SyntaxPatternEmpty rhs =
-      \case
+    match next SyntaxPatternEmpty rhs scrutinee env k =
+      case scrutinee of
         (ValueSyntax (Syntax (Stx _ _ (List [])))) ->
-          eval rhs
+          step $ Down (unCore rhs) env k
         _ -> next
-    match next (SyntaxPatternCons nx x nxs xs) rhs =
-      \case
+    match next (SyntaxPatternCons nx x nxs xs) rhs scrutinee env k =
+      case scrutinee of
         (ValueSyntax (Syntax (Stx scs loc (List (v:vs))))) ->
-          withExtendedEnv nx x (ValueSyntax v) $
-          withExtendedEnv nxs xs (ValueSyntax (Syntax (Stx scs loc (List vs)))) $
-          eval rhs
+          let mkEnv = extend nx x (ValueSyntax v)
+                    . extend nxs xs (ValueSyntax (Syntax (Stx scs loc (List vs))))
+          in step $ Down (unCore rhs) (mkEnv env) k
         _ -> next
-    match next (SyntaxPatternList xs) rhs =
-      \case
+    match next (SyntaxPatternList xs) rhs scrutinee env k =
+      case scrutinee of
         (ValueSyntax (Syntax (Stx _ _ (List vs))))
           | length vs == length xs ->
-            withManyExtendedEnv [(n, x, (ValueSyntax v))
-                                | (n,x) <- xs
-                                | v <- vs] $
-            eval rhs
+            let vals = [ (n, x, ValueSyntax v)
+                       | (n,x) <- xs
+                       | v     <- vs
+                       ]
+            in step $ Down (unCore rhs) (vals `extends` env) k
         _ -> next
-    match _next SyntaxPatternAny rhs =
-      const (eval rhs)
+    match _next SyntaxPatternAny rhs _scrutinee env k =
+      step $ Down (unCore rhs) env k
+
+doDataCase :: SrcLoc -> Value -> [(ConstructorPattern, Core)] -> VEnv -> Kont -> EState
+doDataCase loc v0 [] _e _kont = error $ show (EvalErrorCase loc v0)
+doDataCase loc v0 ((pat, rhs) : ps) env kont =
+  match (doDataCase loc v0 ps env kont) (\newEnv -> step $ Down (unCore rhs) newEnv kont) [(unConstructorPattern pat, v0)]
+  where
+    match
+      :: EState {- ^ Failure continuation -}
+      -> (VEnv -> EState) {- ^ Success continuation, to be used in an extended environment -}
+      -> [(ConstructorPatternF ConstructorPattern, Value)] {- ^ Subpatterns and their scrutinees -}
+      -> EState
+    match _fk sk [] = sk env
+    match fk sk ((CtorPattern ctor subPats, tgt) : more) =
+      case tgt of
+        ValueCtor c args
+          | c == ctor ->
+            if length subPats /= length args
+              then error $ "Type checker bug: wrong number of pattern vars for constructor " ++ show c
+              else match fk sk (zip (map unConstructorPattern subPats) args ++ more)
+        _otherValue -> fk
+    match fk sk ((PatternVar n x, tgt) : more) =
+      match fk (sk . extend n x tgt) more
```

Comment:
@coderabbit could you please find the previous implementation of `doDataCase` and show me the diff?